File: examples/ios_shell_host/CrosswakeShell/TransferCoordinator.swift
import Foundation
import SwiftUI
#if canImport(UniformTypeIdentifiers)
import UniformTypeIdentifiers
#endif

@MainActor
final class TransferCoordinator: ObservableObject {
    enum TransferCommand: String, CaseIterable {
        case transferImport = "transfer.import"
        case transferExport = "transfer.export"
        case transferDownload = "transfer.download"
        case transferUploadPrepare = "transfer.upload.prepare"
    }

    enum TransferState: String, CaseIterable {
        case queued = "queued"
        case preparing = "preparing"
        case transferring = "transferring"
        case awaitingNetwork = "awaiting_network"
        case verifying = "verifying"
        case complete = "complete"
        case failed = "failed"
        case canceled = "canceled"
    }

    struct TransferRecord: Equatable {
        let routeID: String
        let transferID: String
        let command: TransferCommand
        let intent: String
        let source: String?
        let destination: String?
        let state: TransferState
        let detail: String
        let stagedPath: String?
    }

    struct StagedDocument: Equatable {
        let routeID: String
        let transferID: String
        let handle: String
        let stagedPath: String
        let name: String?
        let mimeType: String?
        let nativeType: String?
        let sizeBytes: Int?
    }

    private let routeID: String
    private let declaredTransfers: [ShellManifest.TransferSeam]

    @Published private(set) var transfers: [String: TransferRecord] = [:]
    @Published private(set) var stagedDocuments: [String: StagedDocument] = [:]

    init(routeID: String, declaredTransfers: [ShellManifest.TransferSeam]) {
        self.routeID = routeID
        self.declaredTransfers = declaredTransfers
    }

    func execute(command: String, payload: [String: String], correlationID: String) -> [String: String]? {
        guard let transferCommand = TransferCommand(rawValue: command),
              let transferID = payload["transfer_id"],
              let seam = declaredTransfer(id: transferID, matching: transferCommand) else {
            return nil
        }

        let record = transition(
            transferID: transferID,
            command: transferCommand,
            seam: seam,
            stagedPath: payload["staged_path"],
            correlationID: correlationID
        )

        transfers[transferID] = record

        return [
            "route_id": routeID,
            "transfer_id": transferID,
            "state": record.state.rawValue,
            "intent": seam.intent,
            "detail": record.detail,
            "correlation_id": correlationID
        ]
    }

    func stageCapturedMedia(
        transferID: String,
        localPath: String,
        mediaType: String,
        bytes: Int
    ) -> [String: String]? {
        guard let seam = declaredTransfers.first(where: {
            $0.id == transferID && $0.intent == "upload" && $0.source == "native_capture"
        }) else {
            return nil
        }

        let record = TransferRecord(
            routeID: routeID,
            transferID: transferID,
            command: .transferUploadPrepare,
            intent: seam.intent,
            source: seam.source,
            destination: seam.destination,
            state: .queued,
            detail: "Captured locally. Transfer handoff stays explicit until the route starts upload preparation.",
            stagedPath: localPath
        )

        transfers[transferID] = record

        return [
            "route_id": routeID,
            "transfer_id": transferID,
            "state": record.state.rawValue,
            "intent": seam.intent,
            "staged_path": localPath,
            "media_type": mediaType,
            "bytes": String(bytes)
        ]
    }

    func declaredPickerTransfer(id: String) -> ShellManifest.TransferSeam? {
        declaredTransfers.first(where: { seam in
            seam.id == id && seam.direction == "inbound" && seam.source == "native_picker"
        })
    }

    func markPickerCanceled(transferID: String, correlationID: String) {
        guard let seam = declaredPickerTransfer(id: transferID) else { return }

        transfers[transferID] = TransferRecord(
            routeID: routeID,
            transferID: transferID,
            command: .transferImport,
            intent: seam.intent,
            source: seam.source,
            destination: seam.destination,
            state: .canceled,
            detail: "Picker canceled before any staged copy was handed to the route. [\(correlationID)]",
            stagedPath: nil
        )
    }

    func stagePickedDocument(
        transferID: String,
        sourceURL: URL,
        correlationID: String
    ) throws -> [String: String]? {
        guard let seam = declaredPickerTransfer(id: transferID) else {
            return nil
        }

        let fileManager = FileManager.default
        let stagingDirectory = fileManager.temporaryDirectory
            .appendingPathComponent("crosswake-picker-staging", isDirectory: true)
            .appendingPathComponent(transferID, isDirectory: true)

        try fileManager.createDirectory(at: stagingDirectory, withIntermediateDirectories: true)

        let handle = UUID().uuidString.lowercased()
        let fileExtension = sourceURL.pathExtension.isEmpty ? "" : ".\(sourceURL.pathExtension)"
        let stagedURL = stagingDirectory.appendingPathComponent(handle + fileExtension)
        let accessedSecurityScope = sourceURL.startAccessingSecurityScopedResource()

        defer {
            if accessedSecurityScope {
                sourceURL.stopAccessingSecurityScopedResource()
            }
        }

        if fileManager.fileExists(atPath: stagedURL.path) {
            try fileManager.removeItem(at: stagedURL)
        }

        try fileManager.copyItem(at: sourceURL, to: stagedURL)

        let resourceValues = try stagedURL.resourceValues(forKeys: [.contentTypeKey, .fileSizeKey, .nameKey])
        let nativeType = resourceValues.contentType?.identifier
        let mimeType = resourceValues.contentType?.preferredMIMEType
        let name = resourceValues.name ?? sourceURL.lastPathComponent
        let resolvedName = name.isEmpty ? nil : name
        let sizeBytes = resourceValues.fileSize

        stagedDocuments[handle] = StagedDocument(
            routeID: routeID,
            transferID: transferID,
            handle: handle,
            stagedPath: stagedURL.path,
            name: resolvedName,
            mimeType: mimeType,
            nativeType: nativeType,
            sizeBytes: sizeBytes
        )

        transfers[transferID] = TransferRecord(
            routeID: routeID,
            transferID: transferID,
            command: .transferImport,
            intent: seam.intent,
            source: seam.source,
            destination: seam.destination,
            state: .queued,
            detail: "Picker selection copied into app staging. Transfer verification still owns MIME and integrity truth. [\(correlationID)]",
            stagedPath: stagedURL.path
        )

        var item = ["handle": handle]
        if let resolvedName {
            item["name"] = resolvedName
        }
        if let mimeType, mimeType.isEmpty == false {
            item["mime_type"] = mimeType
        }
        if let sizeBytes {
            item["size_bytes"] = String(sizeBytes)
        }
        if let nativeType, nativeType.isEmpty == false {
            item["native_type"] = nativeType
        }

        return item
    }

    private func declaredTransfer(id: String, matching command: TransferCommand) -> ShellManifest.TransferSeam? {
        declaredTransfers.first(where: { seam in
            seam.id == id && commandMatchesDeclaredIntent(command, intent: seam.intent)
        })
    }

    private func commandMatchesDeclaredIntent(_ command: TransferCommand, intent: String) -> Bool {
        switch (command, intent) {
        case (.transferImport, "import"),
             (.transferExport, "export"),
             (.transferDownload, "download"),
             (.transferUploadPrepare, "upload"):
            return true
        default:
            return false
        }
    }

    private func transition(
        transferID: String,
        command: TransferCommand,
        seam: ShellManifest.TransferSeam,
        stagedPath: String?,
        correlationID: String
    ) -> TransferRecord {
        let state: TransferState
        let detail: String

        switch command {
        case .transferUploadPrepare:
            state = stagedPath == nil ? .failed : .preparing
            detail = stagedPath == nil
                ? "Upload preparation requires an explicitly staged local media file."
                : "Upload preparation acknowledged for staged media. Foreground transfer remains route-local."
        case .transferImport:
            state = .preparing
            detail = "Import preparation acknowledged. Route-local transfer remains explicit."
        case .transferExport:
            state = .queued
            detail = "Export request queued for the active route only."
        case .transferDownload:
            state = .transferring
            detail = "Download started in the foreground for the active route only."
        }

        return TransferRecord(
            routeID: routeID,
            transferID: transferID,
            command: command,
            intent: seam.intent,
            source: seam.source,
            destination: seam.destination,
            state: state,
            detail: detail + " [\(correlationID)]",
            stagedPath: stagedPath
        )
    }
}

File: examples/android_shell_host/app/src/main/java/dev/crosswake/shell/transfer/TransferCoordinator.kt
package dev.crosswake.shell.transfer

import dev.crosswake.shell.ShellManifest

class TransferCoordinator(
    private val routeId: String,
    private val declaredTransfers: List<ShellManifest.TransferSeam>
) {
    data class StagedPickerItem(
        val handle: String,
        val localPath: String,
        val name: String?,
        val mimeType: String?,
        val sizeBytes: Long?,
        val nativeType: String?
    )

    enum class TransferCommand(val wireValue: String) {
        TRANSFER_IMPORT("transfer.import"),
        TRANSFER_EXPORT("transfer.export"),
        TRANSFER_DOWNLOAD("transfer.download"),
        TRANSFER_UPLOAD_PREPARE("transfer.upload.prepare");

        companion object {
            fun fromWireValue(value: String): TransferCommand? = entries.find { it.wireValue == value }
        }
    }

    enum class TransferState(val wireValue: String) {
        QUEUED("queued"),
        PREPARING("preparing"),
        TRANSFERRING("transferring"),
        AWAITING_NETWORK("awaiting_network"),
        VERIFYING("verifying"),
        COMPLETE("complete"),
        FAILED("failed"),
        CANCELED("canceled")
    }

    data class TransferRecord(
        val routeId: String,
        val transferId: String,
        val command: TransferCommand,
        val intent: String,
        val source: String?,
        val destination: String?,
        val state: TransferState,
        val detail: String,
        val stagedPath: String?
    )

    private val transfers = mutableMapOf<String, TransferRecord>()
    private val stagedPickerItems = mutableMapOf<String, List<StagedPickerItem>>()

    fun execute(command: String, payload: Map<String, String>, correlationId: String): Map<String, String>? {
        val transferCommand = TransferCommand.fromWireValue(command) ?: return null
        val transferId = payload["transfer_id"] ?: return null
        val seam = declaredTransfer(transferId, transferCommand) ?: return null

        val record = transition(transferId, transferCommand, seam, payload["staged_path"], correlationId)
        transfers[transferId] = record

        return mapOf(
            "route_id" to routeId,
            "transfer_id" to transferId,
            "state" to record.state.wireValue,
            "intent" to seam.intent,
            "detail" to record.detail,
            "correlation_id" to correlationId
        )
    }

    fun stageCapturedMedia(
        transferId: String,
        localPath: String,
        mediaType: String,
        bytes: Long
    ): Map<String, String>? {
        val seam = declaredTransfers.firstOrNull {
            it.id == transferId && it.intent == "upload" && it.source == "native_capture"
        } ?: return null

        val record = TransferRecord(
            routeId = routeId,
            transferId = transferId,
            command = TransferCommand.TRANSFER_UPLOAD_PREPARE,
            intent = seam.intent,
            source = seam.source,
            destination = seam.destination,
            state = TransferState.QUEUED,
            detail = "Captured locally. Transfer handoff stays explicit until the route starts upload preparation.",
            stagedPath = localPath
        )

        transfers[transferId] = record

        return mapOf(
            "route_id" to routeId,
            "transfer_id" to transferId,
            "state" to record.state.wireValue,
            "intent" to seam.intent,
            "staged_path" to localPath,
            "media_type" to mediaType,
            "bytes" to bytes.toString()
        )
    }

    fun declaredNativePickerTransfer(transferId: String): ShellManifest.TransferSeam? {
        return declaredTransfers.firstOrNull { seam ->
            seam.id == transferId &&
                seam.source == "native_picker" &&
                (seam.intent == "import" || seam.intent == "upload")
        }
    }

    fun stagePickedFiles(
        transferId: String,
        items: List<StagedPickerItem>,
        correlationId: String
    ): Map<String, String>? {
        val seam = declaredNativePickerTransfer(transferId) ?: return null
        val command =
            if (seam.intent == "upload") {
                TransferCommand.TRANSFER_UPLOAD_PREPARE
            } else {
                TransferCommand.TRANSFER_IMPORT
            }

        val record = TransferRecord(
            routeId = routeId,
            transferId = transferId,
            command = command,
            intent = seam.intent,
            source = seam.source,
            destination = seam.destination,
            state = TransferState.PREPARING,
            detail = "Picker content copied into app-controlled staging. Transfer verification remains route-local. [$correlationId]",
            stagedPath = items.firstOrNull()?.localPath
        )

        transfers[transferId] = record
        stagedPickerItems[transferId] = items

        val payload = linkedMapOf(
            "status" to "ok",
            "transfer_id" to transferId
        )

        items.forEachIndexed { index, item ->
            payload["items.$index.handle"] = item.handle
            item.name?.let { payload["items.$index.name"] = it }
            item.mimeType?.let { payload["items.$index.mime_type"] = it }
            item.sizeBytes?.let { payload["items.$index.size_bytes"] = it.toString() }
            item.nativeType?.let { payload["items.$index.native_type"] = it }
        }

        return payload
    }

    fun cancelPickedFiles(transferId: String, detailReason: String = "user_canceled"): Map<String, String>? {
        val seam = declaredNativePickerTransfer(transferId) ?: return null
        val command =
            if (seam.intent == "upload") {
                TransferCommand.TRANSFER_UPLOAD_PREPARE
            } else {
                TransferCommand.TRANSFER_IMPORT
            }

        val record = TransferRecord(
            routeId = routeId,
            transferId = transferId,
            command = command,
            intent = seam.intent,
            source = seam.source,
            destination = seam.destination,
            state = TransferState.CANCELED,
            detail = "Picker canceled before transfer staging began.",
            stagedPath = null
        )

        transfers[transferId] = record

        return mapOf(
            "status" to "canceled",
            "transfer_id" to transferId,
            "detail.reason" to detailReason
        )
    }

    private fun declaredTransfer(id: String, command: TransferCommand): ShellManifest.TransferSeam? {
        return declaredTransfers.firstOrNull { seam ->
            seam.id == id && commandMatchesIntent(command, seam.intent)
        }
    }

    private fun commandMatchesIntent(command: TransferCommand, intent: String): Boolean {
        return when (command to intent) {
            TransferCommand.TRANSFER_IMPORT to "import",
            TransferCommand.TRANSFER_EXPORT to "export",
            TransferCommand.TRANSFER_DOWNLOAD to "download",
            TransferCommand.TRANSFER_UPLOAD_PREPARE to "upload" -> true
            else -> false
        }
    }

    private fun transition(
        transferId: String,
        command: TransferCommand,
        seam: ShellManifest.TransferSeam,
        stagedPath: String?,
        correlationId: String
    ): TransferRecord {
        val state: TransferState
        val detail: String

        when (command) {
            TransferCommand.TRANSFER_UPLOAD_PREPARE -> {
                state = if (stagedPath == null) TransferState.FAILED else TransferState.PREPARING
                detail =
                    if (stagedPath == null) {
                        "Upload preparation requires an explicitly staged local media file."
                    } else {
                        "Upload preparation acknowledged for staged media. Foreground transfer remains route-local."
                    }
            }
            TransferCommand.TRANSFER_IMPORT -> {
                state = TransferState.PREPARING
                detail = "Import preparation acknowledged. Route-local transfer remains explicit."
            }
            TransferCommand.TRANSFER_EXPORT -> {
                state = TransferState.QUEUED
                detail = "Export request queued for the active route only."
            }
            TransferCommand.TRANSFER_DOWNLOAD -> {
                state = TransferState.TRANSFERRING
                detail = "Download started in the foreground for the active route only."
            }
        }

        return TransferRecord(
            routeId = routeId,
            transferId = transferId,
            command = command,
            intent = seam.intent,
            source = seam.source,
            destination = seam.destination,
            state = state,
            detail = "$detail [$correlationId]",
            stagedPath = stagedPath
        )
    }
}
